extensions/azure-extensions/src/main/java/io/druid/storage/azure/AzureDataSegmentPusher.java

```diff
@@ -0,0 +1,165 @@
+  {
+    final File zipOutFile = File.createTempFile(TEMP_DIR, AzureStorageDruidModule.INDEX_ZIP_FILE_NAME);
+    CompressionUtils.zip(indexFilesDir, zipOutFile);
+
+    return zipOutFile;
+
+  }
+
+  public File createSegmentDescriptorFile(final ObjectMapper jsonMapper, final DataSegment segment) throws
+                                                                                                           IOException
+  {
+    File descriptorFile = File.createTempFile(TEMP_DIR, AzureStorageDruidModule.DESCRIPTOR_FILE_NAME);
+    try (FileOutputStream stream = new FileOutputStream(descriptorFile)) {
+      stream.write(jsonMapper.writeValueAsBytes(segment));
+    }
+
+    return descriptorFile;
+  }
+
+  public Map<String, String> getAzurePaths(final DataSegment segment)
+  {
+    final String storageDir = DataSegmentPusherUtil.getStorageDir(segment);
+
+    return ImmutableMap.of(
+        "storage", storageDir,
+        "index", String.format("%s/%s", storageDir, AzureStorageDruidModule.INDEX_ZIP_FILE_NAME),
+        "descriptor", String.format("%s/%s", storageDir, AzureStorageDruidModule.DESCRIPTOR_FILE_NAME)
+    );
+
+  }
+
+  public <T> T retryAzureOperation(Callable<T> f, int maxTries) throws Exception
+  {
+    return RetryUtils.retry(f, AzureUtils.AZURE_RETRY, maxTries);
+  }
+
+  @Override
+  public DataSegment push(final File indexFilesDir, final DataSegment segment) throws IOException
+  {
+
+    log.info("Uploading [%s] to Azure.", indexFilesDir);
+
+    final int version = SegmentUtils.getVersionFromDir(indexFilesDir);
+    final File compressedSegmentData = createCompressedSegmentDataFile(indexFilesDir);
+    final File descriptorFile = createSegmentDescriptorFile(jsonMapper, segment);
+    final Map<String, String> azurePaths = getAzurePaths(segment);
+
+    try {
+      return retryAzureOperation(
+          new Callable<DataSegment>()
+          {
+            @Override
+            public DataSegment call() throws Exception
+            {
+              return uploadDataSegment(segment, version, compressedSegmentData, descriptorFile, azurePaths);
+            }
+          },
+          azureAccountConfig.getMaxTries()
+      );
+    }
+    catch (Exception e) {
+      throw Throwables.propagate(e);
+    }
+  }
+
+  public void uploadThenDelete(final File file, final String azurePath)
+      throws StorageException, IOException, URISyntaxException
+  {
+    azureStorageContainer.uploadBlob(file, azurePath);
+    log.info("Deleting file [%s]", file);
+    file.delete();
+  }
+
+  public DataSegment uploadDataSegment(
+      DataSegment segment,
+      final int version,
+      final File compressedSegmentData,
+      final File descriptorFile,
+      final Map<String, String> azurePaths
+  )
+      throws StorageException, IOException, URISyntaxException
+  {
+    uploadThenDelete(compressedSegmentData, azurePaths.get("index"));
+    uploadThenDelete(descriptorFile, azurePaths.get("descriptor"));
```

Comment:
I wonder if you really need to create those temp files, or can use Azure BlobOutputStream as is done in HDFSDeepStorage.